File: libraries/android/sdk/src/main/java/com/voidhash/sdk/PurchaseOrchestrator.kt
package com.voidhash.sdk

import com.voidhash.core.billing.BillingBuyItemParams
import com.voidhash.core.billing.BillingProductType
import com.voidhash.sdk.api.DevelopmentPurchaseRequest
import com.voidhash.sdk.api.SyncTransactionRequest
import com.voidhash.sdk.api.TransactionSyncVerdict
import com.voidhash.sdk.api.VoidhashApiClient
import com.voidhash.sdk.billing.BillingEnginePort
import com.voidhash.sdk.billing.VoidhashProduct
import com.voidhash.sdk.billing.VoidhashTransaction
import com.voidhash.sdk.billing.mapBillingProductToProduct
import com.voidhash.sdk.billing.mapBillingPurchaseToTransaction
import com.voidhash.sdk.billing.mapDevelopmentPurchaseToTransaction
import com.voidhash.sdk.cache.CacheManager
import com.voidhash.sdk.identity.AccountToken
import com.voidhash.sdk.identity.IdentityStore
import com.voidhash.sdk.network.VoidhashCircuitOpenException
import com.voidhash.sdk.network.VoidhashOutboundPausedException
import com.voidhash.sdk.schema.RuntimeProductDefinition
import com.voidhash.sdk.schema.RuntimeSchema
import com.voidhash.sdk.transactions.TransactionOutbox
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONObject

private const val PROCESSED_TRANSACTION_TTL_MS = 1000L * 60 * 30

private class InFlightTransaction {
    val deferred = CompletableDeferred<Boolean>()
    var ownerClaimed = false
    var storeFinalizationPending = false
}

/** The outcome of claiming a transaction: either this caller owns the run or it joins one. */
private class TransactionClaim(val entry: InFlightTransaction, val isOwner: Boolean)

/**
 * Owns the transaction lifecycle: buying through Play Billing, syncing to the
 * backend, and finishing (acknowledging or consuming) with the store.
 *
 * Processing is deduplicated per transaction both in memory (concurrent callers
 * coalesce onto one run) and across restarts (a 30-minute processed-transaction
 * cache entry recording `{backendAccepted, storeFinalized}`). Observer mode
 * (`readOnly`) syncs but never finishes with the store — except for a purchase
 * this SDK started, whose ownership is pinned when the flow begins so a
 * mid-flight `setReadOnly(true)` can't strand the transaction.
 */
class PurchaseOrchestrator(
    private val billing: BillingEnginePort,
    private val apiClient: VoidhashApiClient,
    private val cacheManager: CacheManager,
    private val identityStore: IdentityStore,
    private val readOnlyProvider: () -> Boolean = { false },
    /** When true purchases run against the mock store instead of Play Billing. */
    private val developmentMode: Boolean = false,
    private val onPersonRefresh: suspend () -> Unit = {},
    /** Durable receipt store; a receipt is recorded here before it is ever sent. */
    private val outbox: TransactionOutbox? = null,
    private val onWarning: (String) -> Unit = {},
) {
    private val inFlightTransactions = mutableMapOf<String, InFlightTransaction>()

    // The observer callback processes transactions on its own coroutine, so the
    // registry and the entry flags are reached from several threads; both the
    // lookups and the check-then-act flag transitions run under this mutex.
    private val inFlightMutex = Mutex()

    /** Queries the store for every product configured in [schema]. */
    suspend fun getProducts(schema: RuntimeSchema): List<VoidhashProduct> {
        if (developmentMode) {
            return schema.products.values.mapNotNull { definition ->
                val configuration = definition.providers.development ?: return@mapNotNull null
                VoidhashProduct(
                    id = configuration.productId,
                    slug = definition.slug,
                    name = definition.name,
                    description = "Development purchase",
                    displayName = definition.name,
                    displayPrice = "$" + "%.2f".format(configuration.price),
                    price = configuration.price,
                    currency = configuration.currencyCode,
                    type = if (definition.type == "subscription") "subs" else "inapp",
                    billingPeriod = configuration.period,
                    googlePlayOfferToken = null,
                )
            }
        }

        val definitionsByProductId = schema.products.values
            .mapNotNull { definition ->
                definition.providers.googlePlay?.productId?.let { it to definition }
            }
            .toMap()

        if (definitionsByProductId.isEmpty()) {
            return emptyList()
        }

        val productIds = definitionsByProductId.keys.toTypedArray()
        val details = billing.getItemsByType(BillingProductType.INAPP, productIds) +
            billing.getItemsByType(BillingProductType.SUBS, productIds)

        return details.map { detail ->
            mapBillingProductToProduct(definitionsByProductId[detail.id], detail)
        }
    }

    /**
     * Buys [product], syncs the resulting transaction and refreshes the person.
     */
    internal suspend fun purchase(
        product: VoidhashProduct,
        schema: RuntimeSchema,
    ): VoidhashTransaction {
        // Pinned at purchase start: a purchase this SDK owns must still be
        // finished with the store even if the app flips to observer mode while
        // the store sheet is open.
        val readOnlyAtPurchaseStart = readOnlyProvider()
        val distinctId = identityStore.getDistinctId()

        val type = if (product.isSubscription) BillingProductType.SUBS else BillingProductType.INAPP
        if (!developmentMode && type == BillingProductType.SUBS && product.googlePlayOfferToken == null) {
            throw VoidhashException(
                "PURCHASE_FAILED",
                "Google Play subscription has no configured offer token",
            )
        }

        val purchases = billing.buyItemByType(
            BillingBuyItemParams(
                type = type,
                skuArr = arrayOf(product.id),
                obfuscatedAccountId = AccountToken.derive(distinctId),
                offerTokenArr = product.googlePlayOfferToken?.let { arrayOf(it) },
                isOfferPersonalized = false,
            ),
        )

        val purchase = purchases.firstOrNull()
            ?: throw VoidhashException("PURCHASE_FAILED", "No purchase returned from Google Billing")

        val transaction = if (developmentMode) {
            mapDevelopmentPurchaseToTransaction(
                productId = product.id,
                devTransactionId = purchase.orderId ?: purchase.purchaseToken,
                purchaseDate = purchase.purchaseTime,
                quantity = 1,
            )
        } else {
            mapBillingPurchaseToTransaction(purchase)
        }
        when (transaction.purchaseState) {
            "purchased" -> Unit
            "pending" -> throw VoidhashException("PURCHASE_PENDING", "The payment was deferred")
            else -> throw VoidhashException(
                "PURCHASE_UNKNOWN_RESULT",
                "Google Billing returned an unknown state",
            )
        }

        if (processTransaction(transaction, schema, readOnlyAtPurchaseStart)) {
            refreshPerson("a purchase")
        }
        return transaction
    }

    /** Restores purchases: reconciles everything the store still reports, then refreshes. */
    suspend fun restorePurchases(schema: RuntimeSchema) {
        reconcileObservedTransactions(schema)
        refreshPerson("a restore")
    }

    /**
     * Processes every transaction the store still reports. One-time consumables
     * are skipped: consuming them here would silently burn an entitlement the
     * app has already granted. When [deferStoreFinalization] is true, receipts
     * are synced but left unfinished until a schema can classify the product.
     */
    suspend fun reconcileObservedTransactions(
        schema: RuntimeSchema,
        deferStoreFinalization: Boolean = false,
    ) {
        val purchases = billing.getAvailableItemsByType(BillingProductType.INAPP) +
            billing.getAvailableItemsByType(BillingProductType.SUBS)

        val observedByKey = linkedMapOf<String, VoidhashTransaction>()
        for (purchase in purchases) {
            val transaction = mapBillingPurchaseToTransaction(purchase)
            observedByKey[transaction.processingKey] = transaction
        }

        val failures = mutableListOf<String>()
        for (transaction in observedByKey.values) {
            if (resolveProductDefinition(transaction, schema)?.type == "one-time-consumable") {
                continue
            }
            try {
                processTransaction(
                    transaction,
                    schema,
                    if (deferStoreFinalization) true else null,
                )
            } catch (error: CancellationException) {
                throw error
            } catch (error: Throwable) {
                failures.add(transaction.transactionId)
                onWarning("Failed to process observed transaction ${transaction.transactionId}: ${error.message}")
            }
        }

        if (failures.isNotEmpty()) {
            throw VoidhashException(
                "RECONCILE_TRANSACTIONS_FAILED",
                "Failed to restore ${failures.size} transactions",
            )
        }
    }

    /**
     * Processes one observed transaction and refreshes the person snapshot.
     * [deferStoreFinalization] records and syncs the receipt without finishing
     * it in the store until product metadata is available.
     */
    suspend fun processObservedTransaction(
        transaction: VoidhashTransaction,
        schema: RuntimeSchema,
        deferStoreFinalization: Boolean = false,
    ) {
        if (
            processTransaction(
                transaction,
                schema,
                if (deferStoreFinalization) true else null,
            )
        ) {
            refreshPerson("an observed transaction")
        }
    }

    /**
     * Syncs [transaction] to the backend and — unless the SDK is in observer
     * mode — finishes it with the store.
     *
     * [readOnlyOverride] pins the ownership decision for a purchase this SDK
     * started; every other caller passes `null` and reads the live flag at the
     * moment the decision is made.
     *
     * Returns `true` only after the backend explicitly accepts the receipt. A
     * `false` result leaves the receipt queued and the store transaction unfinished.
     */
    suspend fun processTransaction(
        transaction: VoidhashTransaction,
        schema: RuntimeSchema,
        readOnlyOverride: Boolean?,
    ): Boolean {
        if (transaction.purchaseState != "purchased") {
            return false
        }

        val processingKey = transaction.processingKey
        val processedCacheKey = "processed-transaction:$processingKey"

        val claim = claimTransaction(processingKey, readOnlyOverride)
        if (!claim.isOwner) {
            // Rethrows the owner's failure: a joiner must never read a failed —
            // or cancelled — run as a synced, finished transaction.
            val accepted = claim.entry.deferred.await()
            if (accepted && readOnlyOverride == false && takeStoreFinalizationPending(claim.entry)) {
                finalizeWithStore(transaction, schema, processedCacheKey)
            }
            return accepted
        }

        try {
            val accepted = runTransaction(
                transaction,
                schema,
                readOnlyOverride,
                claim.entry,
                processedCacheKey,
                processingKey,
            )
            claim.entry.deferred.complete(accepted)
            return accepted
        } catch (error: Throwable) {
            claim.entry.deferred.completeExceptionally(error)
            throw error
        } finally {
            withContext(NonCancellable) {
                inFlightMutex.withLock { inFlightTransactions.remove(processingKey) }
            }
        }
    }

    private suspend fun claimTransaction(
        processingKey: String,
        readOnlyOverride: Boolean?,
    ): TransactionClaim = inFlightMutex.withLock {
        val existing = inFlightTransactions[processingKey]
        if (existing != null) {
            if (readOnlyOverride == false) {
                existing.ownerClaimed = true
            }
            return@withLock TransactionClaim(existing, isOwner = false)
        }

        val entry = InFlightTransaction()
        entry.ownerClaimed = readOnlyOverride == false
        inFlightTransactions[processingKey] = entry
        TransactionClaim(entry, isOwner = true)
    }

    private suspend fun takeStoreFinalizationPending(entry: InFlightTransaction): Boolean =
        inFlightMutex.withLock {
            val pending = entry.storeFinalizationPending
            entry.storeFinalizationPending = false
            pending
        }

    private suspend fun runTransaction(
        transaction: VoidhashTransaction,
        schema: RuntimeSchema,
        readOnlyOverride: Boolean?,
        entry: InFlightTransaction,
        processedCacheKey: String,
        processingKey: String,
    ): Boolean {
        val cachedState = readProcessedState(processedCacheKey)
        if (cachedState?.storeFinalized == true) {
            return true
        }

        if (!transaction.isDevelopment && transaction.purchaseToken.isNullOrEmpty()) {
            onWarning(
                "Skipping observed Android transaction without purchase token ${transaction.transactionId}",
            )
            return false
        }

        if (cachedState?.backendAccepted != true) {
            val distinctId = identityStore.getDistinctId()
            if (transaction.isDevelopment) {
                val accepted = try {
                    apiClient.developmentPurchase(
                        distinctId,
                        DevelopmentPurchaseRequest(
                            devTransactionId = transaction.transactionId,
                            productSlug = resolveProductSlug(transaction, schema),
                            purchaseDate = transaction.purchaseDate,
                            quantity = transaction.quantity,
                        ),
                    )
                } catch (error: CancellationException) {
                    throw error
                } catch (error: Throwable) {
                    if (!isDeferredSyncFailure(error)) throw error
                    onWarning(
                        "Deferred development transaction ${transaction.transactionId}: ${error.message}",
                    )
                    return false
                }
                if (!accepted) {
                    onWarning("Backend did not accept transaction ${transaction.transactionId}")
                    return false
                }
            } else {
                val request = SyncTransactionRequest(
                    appAccountToken = transaction.appAccountToken,
                    providerProductId = transaction.productId,
                    productSlug = resolveProductSlug(transaction, schema),
                    purchaseDate = transaction.purchaseDate,
                    purchaseToken = transaction.purchaseToken ?: "",
                    quantity = transaction.quantity,
                    receipt = transaction.receipt,
                    transactionId = transaction.transactionId,
                )
                // Recorded before the request so a crash, a kill, or an outage between here
                // and the backend's answer cannot lose a purchase the user already paid for.
                outbox?.enqueue(processingKey, distinctId, request)
                val verdict = try {
                    apiClient.syncTransactionVerdict(distinctId, request)
                } catch (error: CancellationException) {
                    throw error
                } catch (error: Throwable) {
                    if (!isDeferredSyncFailure(error)) throw error
                    onWarning(
                        "Deferred transaction ${transaction.transactionId}: ${error.message}",
                    )
                    outbox?.postpone(processingKey)
                    return false
                }
                when (verdict) {
                    TransactionSyncVerdict.ACCEPTED -> outbox?.acknowledge(processingKey)
                    TransactionSyncVerdict.REJECTED,
                    TransactionSyncVerdict.INDETERMINATE,
                    -> {
                        onWarning("Backend did not accept transaction ${transaction.transactionId}")
                        outbox?.postpone(processingKey)
                        return false
                    }
                }
            }

            writeProcessedState(
                processedCacheKey,
                backendAccepted = true,
                storeFinalized = transaction.isAcknowledged,
            )
        }

        val deferFinalization = inFlightMutex.withLock {
            if (!entry.ownerClaimed && (readOnlyOverride ?: readOnlyProvider())) {
                entry.storeFinalizationPending = true
                true
            } else {
                false
            }
        }
        if (deferFinalization) {
            return true
        }

        finalizeWithStore(transaction, schema, processedCacheKey)
        return true
    }

    private fun isDeferredSyncFailure(error: Throwable): Boolean = when (error) {
        is VoidhashNetworkException -> true
        is VoidhashCircuitOpenException -> true
        is VoidhashOutboundPausedException -> true
        is VoidhashApiException -> true
        else -> false
    }

    /**
     * Refreshes the person snapshot after a synced transaction. A refresh is a
     * read-side convenience: a network blip must not fail a purchase the store
     * and the backend have already accepted.
     */
    private suspend fun refreshPerson(context: String) {
        try {
            onPersonRefresh()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            onWarning("Failed to refresh the person after $context: ${error.message}")
        }
    }

    private suspend fun finalizeWithStore(
        transaction: VoidhashTransaction,
        schema: RuntimeSchema,
        processedCacheKey: String,
    ) {
        if (!transaction.isAcknowledged) {
            val token = transaction.purchaseToken
                ?: throw VoidhashException(
                    "BILLING_ERROR",
                    "Purchase token is required for acknowledgment",
                )

            val result = if (resolveProductDefinition(transaction, schema)?.type == "one-time-consumable") {
                billing.consumeProduct(token)
            } else {
                billing.acknowledgePurchase(token)
            }

            if (result.responseCode != 0.0) {
                throw VoidhashException("BILLING_ERROR", result.message)
            }
        }

        writeProcessedState(processedCacheKey, backendAccepted = true, storeFinalized = true)
    }

    private fun resolveProductDefinition(
        transaction: VoidhashTransaction,
        schema: RuntimeSchema,
    ): RuntimeProductDefinition? = schema.products.values.firstOrNull { definition ->
        definition.slug == transaction.productId ||
            definition.providers.googlePlay?.productId == transaction.productId ||
            definition.providers.development?.productId == transaction.productId
    }

    private fun resolveProductSlug(transaction: VoidhashTransaction, schema: RuntimeSchema): String =
        resolveProductDefinition(transaction, schema)?.slug ?: transaction.productId

    private fun readProcessedState(cacheKey: String): ProcessedTransactionState? {
        val cached = cacheManager.getObject(cacheKey) ?: return null
        // An expired marker is a miss: the cache serves expired entries for offline reads,
        // but a processed-transaction record past its lifetime must not stop a receipt
        // from syncing again.
        if (cached.isExpired) return null
        return ProcessedTransactionState(
            backendAccepted = cached.value.optBoolean("backendAccepted"),
            storeFinalized = cached.value.optBoolean("storeFinalized"),
        )
    }

    private fun writeProcessedState(
        cacheKey: String,
        backendAccepted: Boolean,
        storeFinalized: Boolean,
    ) {
        cacheManager.set(
            cacheKey,
            JSONObject()
                .put("backendAccepted", backendAccepted)
                .put("storeFinalized", storeFinalized),
            ttlMs = PROCESSED_TRANSACTION_TTL_MS,
        )
    }

    private data class ProcessedTransactionState(
        val backendAccepted: Boolean,
        val storeFinalized: Boolean,
    )
}

File: libraries/ios/Sources/Voidhash/Transactions/PurchaseOrchestrator.swift
import Foundation
import VoidhashCore

/// Backend sync surface used by ``PurchaseOrchestrator``; implemented by ``VoidhashApiClient``.
public protocol TransactionSyncing: Sendable {
    /// `POST /api/v1/sdk/sync-transaction`
    func syncTransaction(headers: [String: String], body: SdkSyncTransactionBody) async throws
        -> SdkSyncTransactionResponse
}

extension VoidhashApiClient: TransactionSyncing {}

/// Development-gateway sync surface; implemented by ``VoidhashApiClient``.
public protocol DevelopmentPurchasing: Sendable {
    /// `POST /api/v1/sdk/development/purchase`
    func developmentPurchase(headers: [String: String], body: SdkDevelopmentPurchaseBody)
        async throws
}

extension VoidhashApiClient: DevelopmentPurchasing {}

/// Raised when one or more observed transactions could not be reconciled.
public struct ReconcileTransactionsError: Error, Sendable, CustomStringConvertible, LocalizedError {
    /// Stable error code.
    public let code = "RECONCILE_TRANSACTIONS_FAILED"
    /// Human readable message.
    public let message: String
    /// Identifiers of the transactions that failed.
    public let transactionIds: [String]

    public init(transactionIds: [String]) {
        self.transactionIds = transactionIds
        self.message = "Failed to restore \(transactionIds.count) transactions"
    }

    public var description: String {
        return "\(code): \(message)"
    }

    public var errorDescription: String? {
        return description
    }

    public var localizedDescription: String {
        return description
    }
}

/// Owns the transaction lifecycle: deduplicated backend sync, store finalization and
/// reconciliation of transactions observed outside a purchase this SDK started.
///
/// Ported from `src/core/transactions/transaction-service.ts`; the in-flight map coalesces
/// concurrent processing of the same transaction and the processed-transaction cache
/// (`{backendAccepted, storeFinalized}`, 30 minute TTL) catches duplicates across launches.
public actor PurchaseOrchestrator {
    /// Lifetime of a processed-transaction cache entry.
    public static let processedTransactionTtlMilliseconds: Double = 1000 * 60 * 30

    struct TransactionProcessingState: Codable, Sendable, Equatable {
        let backendAccepted: Bool
        let storeFinalized: Bool
    }

    /// Coalescing entry for one transaction being processed right now.
    ///
    /// `ownerClaimed` is the strongest claim any joined caller made: a purchase this SDK started
    /// pins owner mode and must win over the live read-only flag. `storeFinalizationPending`
    /// records that the running processing synced the transaction but skipped the store finish
    /// because it read read-only mode, so an owner caller that joined too late can finish it
    /// afterwards without re-syncing.
    // Only ever mutated from inside the actor; the `@unchecked` conformance exists so it can be
    // captured by the actor-isolated processing task.
    private final class InFlightTransaction: @unchecked Sendable {
        var ownerClaimed: Bool
        var storeFinalizationPending = false
        var task: Task<Void, any Error>?

        init(ownerClaimed: Bool) {
            self.ownerClaimed = ownerClaimed
        }
    }

    private let engine: any StoreKitEngineProtocol
    private let api: any TransactionSyncing
    private let developmentApi: (any DevelopmentPurchasing)?
    /// When true purchases run against the mock store and are recorded through the
    /// development gateway instead of `sync-transaction`.
    private let isDevelopmentMode: Bool
    private let cacheManager: CacheManager
    private let headersProvider: @Sendable () async -> [String: String]
    private let distinctIdProvider: @Sendable () async -> String
    private let isReadOnly: @Sendable () -> Bool
    private let refreshPerson: @Sendable () async -> Void

    private var inFlightTransactions: [String: InFlightTransaction] = [:]

    /// - Parameters:
    ///   - engine: Store engine used to buy and finish transactions.
    ///   - api: Backend sync client.
    ///   - developmentApi: Development-gateway client; required in development mode.
    ///   - isDevelopmentMode: Routes purchases through the mock store and dev endpoint.
    ///   - cacheManager: Cache holding the processed-transaction states.
    ///   - headersProvider: Builds the common SDK headers for the current identity.
    ///   - distinctIdProvider: Returns the current distinct id.
    ///   - isReadOnly: Reads the live read-only (observer mode) flag.
    ///   - refreshPerson: Refreshes the cached person snapshot after a purchase or restore.
    public init(
        engine: any StoreKitEngineProtocol,
        api: any TransactionSyncing,
        developmentApi: (any DevelopmentPurchasing)? = nil,
        isDevelopmentMode: Bool = false,
        cacheManager: CacheManager,
        headersProvider: @escaping @Sendable () async -> [String: String],
        distinctIdProvider: @escaping @Sendable () async -> String,
        isReadOnly: @escaping @Sendable () -> Bool,
        refreshPerson: @escaping @Sendable () async -> Void
    ) {
        self.engine = engine
        self.api = api
        self.developmentApi = developmentApi
        self.isDevelopmentMode = isDevelopmentMode
        self.cacheManager = cacheManager
        self.headersProvider = headersProvider
        self.distinctIdProvider = distinctIdProvider
        self.isReadOnly = isReadOnly
        self.refreshPerson = refreshPerson
    }

    /// Buys `product`, syncs the resulting transaction and finishes it with the store.
    ///
    /// The read-only flag is pinned when the purchase starts: a purchase this SDK owns is still
    /// finished even if the host app flips to observer mode while the store sheet is open.
    public func purchase(product: VoidhashProduct, schema: RuntimeSchema) async throws {
        let readOnlyAtPurchaseStart = isReadOnly()
        let distinctId = await distinctIdProvider()
        let storeTransaction = try await engine.buyProduct(
            sku: product.id,
            appAccountToken: AccountToken.derive(distinctId: distinctId),
            quantity: 1
        )
        // Dev purchases have nothing to acknowledge — the backend records them directly.
        let transaction = VoidhashTransaction(
            storeTransaction: storeTransaction,
            isDevelopment: isDevelopmentMode,
            isAcknowledgedOverride: isDevelopmentMode
        )
        try await processTransaction(
            transaction, schema: schema, readOnlyOverride: readOnlyAtPurchaseStart)
        await refreshPerson()
    }

    /// Reconciles every observed transaction and refreshes the person snapshot.
    public func restorePurchases(schema: RuntimeSchema) async throws {
        try await reconcileObservedTransactions(schema: schema)
        await refreshPerson()
    }

    /// Syncs pending and still-owned transactions the SDK has not processed yet.
    ///
    /// One-time consumables are skipped: they are finished by the purchase flow that bought them
    /// and must not be re-synced from the entitlement stream. A store read that fails fails the
    /// whole reconciliation — reporting "restored nothing" off an unreadable store would hide a
    /// customer's purchases.
    public func reconcileObservedTransactions(schema: RuntimeSchema) async throws {
        let pending = try engine.getPendingTransactions()
        let purchased = try await engine.getPurchasedItems(onlyIncludeActiveItems: true)

        var observedByKey: [String: VoidhashTransaction] = [:]
        var orderedKeys: [String] = []
        for storeTransaction in pending + purchased {
            let transaction = VoidhashTransaction(storeTransaction: storeTransaction)
            let key = PurchaseOrchestrator.processingKey(transaction)
            if observedByKey[key] == nil {
                orderedKeys.append(key)
            }
            observedByKey[key] = transaction
        }

        var failedTransactionIds: [String] = []
        for key in orderedKeys {
            guard let transaction = observedByKey[key] else {
                continue
            }
            let definition = PurchaseOrchestrator.resolveProductDefinition(
                transaction, products: schema.products)
            if definition?.type == "one-time-consumable" {
                continue
            }
            do {
                try await processTransaction(transaction, schema: schema)
            } catch {
                failedTransactionIds.append(transaction.transactionId)
            }
        }

        if !failedTransactionIds.isEmpty {
            throw ReconcileTransactionsError(transactionIds: failedTransactionIds)
        }
    }

    /// Processes a transaction reported by the store observer and refreshes the person snapshot.
    public func processObservedTransaction(
        _ transaction: VoidhashTransaction, schema: RuntimeSchema
    ) async throws {
        try await processTransaction(transaction, schema: schema)
        await refreshPerson()
    }

    /// Syncs one transaction to the backend and — unless the SDK is read-only — finishes it.
    ///
    /// - Parameter readOnlyOverride: Pins the ownership decision for a purchase this SDK started.
    ///   Every other caller omits it and reads the live flag when the decision is made.
    public func processTransaction(
        _ transaction: VoidhashTransaction,
        schema: RuntimeSchema,
        readOnlyOverride: Bool? = nil
    ) async throws {
        guard transaction.purchaseState == .purchased else {
            return
        }

        let key = PurchaseOrchestrator.processingKey(transaction)
        let processedCacheKey = PurchaseOrchestrator.processedCacheKey(key)

        if let existing = inFlightTransactions[key] {
            guard readOnlyOverride == false else {
                try await existing.task?.value
                return
            }

            existing.ownerClaimed = true
            try await existing.task?.value
            try await finalizeSkippedStoreFinalization(
                transaction, schema: schema, entry: existing, processedCacheKey: processedCacheKey)
            return
        }

        let entry = InFlightTransaction(ownerClaimed: readOnlyOverride == false)
        inFlightTransactions[key] = entry
        let task = Task { [entry] in
            try await self.runProcessing(
                transaction,
                schema: schema,
                entry: entry,
                readOnlyOverride: readOnlyOverride,
                processedCacheKey: processedCacheKey
            )
        }
        entry.task = task

        defer { inFlightTransactions[key] = nil }
        try await task.value
    }

    private func runProcessing(
        _ transaction: VoidhashTransaction,
        schema: RuntimeSchema,
        entry: InFlightTransaction,
        readOnlyOverride: Bool?,
        processedCacheKey: String
    ) async throws {
        let cachedState = await cacheManager.get(
            processedCacheKey, as: TransactionProcessingState.self)?.value

        if cachedState?.storeFinalized == true {
            return
        }

        if cachedState?.backendAccepted != true {
            var headers = await headersProvider()
            headers["x-distinct-id"] = await distinctIdProvider()

            if transaction.isDevelopment {
                guard let developmentApi else {
                    throw VoidhashStoreError(
                        code: "CONFIGURATION_MISSING",
                        message: "Development purchases need a development gateway client")
                }
                try await developmentApi.developmentPurchase(
                    headers: headers,
                    body: SdkDevelopmentPurchaseBody(
                        devTransactionId: transaction.transactionId,
                        productSlug: PurchaseOrchestrator.resolveProductSlug(
                            transaction, products: schema.products),
                        purchaseDate: transaction.purchaseDate,
                        quantity: transaction.quantity
                    ))
            } else {
                let response = try await api.syncTransaction(
                    headers: headers,
                    body: PurchaseOrchestrator.syncPayload(transaction, products: schema.products)
                )

                // Fail-safe: a backend that answers `accepted: false` did not record the
                // purchase, so the transaction stays unfinished (the store re-delivers it)
                // and nothing is cached.
                guard response.accepted else {
                    throw VoidhashStoreError.transactionVerificationRejected(
                        transactionId: transaction.transactionId)
                }
            }

            await cacheManager.set(
                processedCacheKey,
                value: TransactionProcessingState(
                    backendAccepted: true, storeFinalized: transaction.isAcknowledged),
                ttl: PurchaseOrchestrator.processedTransactionTtlMilliseconds
            )
        }

        if !entry.ownerClaimed && (readOnlyOverride ?? isReadOnly()) {
            entry.storeFinalizationPending = true
            return
        }

        try await finalizeWithStore(transaction, processedCacheKey: processedCacheKey)
    }

    /// Runs the store finish that an already-completed processing skipped because it read
    /// read-only mode. The pending flag is cleared in the same isolated step that claims it, so
    /// concurrent owner callers can never both finish the transaction.
    private func finalizeSkippedStoreFinalization(
        _ transaction: VoidhashTransaction,
        schema: RuntimeSchema,
        entry: InFlightTransaction,
        processedCacheKey: String
    ) async throws {
        guard entry.storeFinalizationPending else {
            return
        }
        entry.storeFinalizationPending = false
        try await finalizeWithStore(transaction, processedCacheKey: processedCacheKey)
    }

    private func finalizeWithStore(
        _ transaction: VoidhashTransaction, processedCacheKey: String
    ) async throws {
        if !transaction.isAcknowledged {
            try await engine.finishTransaction(transactionId: transaction.transactionId)
        }

        await cacheManager.set(
            processedCacheKey,
            value: TransactionProcessingState(backendAccepted: true, storeFinalized: true),
            ttl: PurchaseOrchestrator.processedTransactionTtlMilliseconds
        )
    }

    static func processingKey(_ transaction: VoidhashTransaction) -> String {
        return "ios:\(transaction.transactionId):\(formatNumber(transaction.purchaseDate))"
    }

    static func processedCacheKey(_ processingKey: String) -> String {
        return "processed-transaction:\(processingKey)"
    }

    /// Resolves the dashboard product definition a store transaction belongs to.
    static func resolveProductDefinition(
        _ transaction: VoidhashTransaction,
        products: [String: RuntimeProductDefinition]
    ) -> RuntimeProductDefinition? {
        return products.values.first { definition in
            if definition.slug == transaction.productId {
                return true
            }
            if definition.configuration.providers.appleAppStore?.productId
                == transaction.productId
            {
                return true
            }
            return definition.configuration.providers.development?.productId
                == transaction.productId
        }
    }

    static func resolveProductSlug(
        _ transaction: VoidhashTransaction,
        products: [String: RuntimeProductDefinition]
    ) -> String {
        return resolveProductDefinition(transaction, products: products)?.slug
            ?? transaction.productId
    }

    /// Builds the iOS `sync-transaction` payload, mirroring `transaction-service.ts`.
    static func syncPayload(
        _ transaction: VoidhashTransaction,
        products: [String: RuntimeProductDefinition]
    ) -> SdkSyncTransactionBody {
        return SdkSyncTransactionBody(
            appAccountToken: transaction.appAccountToken,
            platform: "ios",
            providerProductId: transaction.productId,
            productSlug: resolveProductSlug(transaction, products: products),
            purchaseDate: transaction.purchaseDate,
            purchaseToken: nil,
            quantity: transaction.quantity,
            receipt: transaction.receipt,
            transactionId: transaction.transactionId
        )
    }

    // The processing key is compared against the TypeScript SDK's, which formats whole numbers
    // without a fractional part.
    private static func formatNumber(_ value: Double) -> String {
        if value.rounded() == value && abs(value) < 1e15 {
            return String(Int64(value))
        }
        return String(value)
    }
}
